bootstore/src/schemes/v0/fsm.rs

```diff
@@ -0,0 +1,820 @@
+
+//! V0 protocol state machine
+//!
+//! This state machine is entirely synchronous. It performs actions and returns
+//! results. This is where the bulk of the protocol logic lives. It's written
+//! this way to enable easy testing and auditing.
+
+use super::request_manager::ShareAcks;
+use super::{
+    create_pkgs, Config, Envelope, LearnedSharePkg, Msg, MsgError, Request,
+    RequestManager, RequestType, Response, ResponseType, Share, SharePkg,
+    Shares, TrackableRequest,
+};
+use crate::trust_quorum::{RackSecret, TrustQuorumError};
+use crate::Sha3_256Digest;
+use secrecy::ExposeSecret;
+use serde::{Deserialize, Serialize};
+use sha3::{Digest, Sha3_256};
+use sled_hardware::Baseboard;
+use std::collections::{BTreeMap, BTreeSet};
+use std::fmt::Debug;
+use std::time::Instant;
+use thiserror::Error;
+use uuid::Uuid;
+
+// An index into an encrypted share
+#[derive(
+    Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize,
+)]
+pub struct ShareIdx(pub usize);
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+pub enum State {
+    Uninitialized,
+    InitialMember {
+        pkg: SharePkg,
+
+        /// Shares given to other sleds. We mark them as used so that we don't
+        /// hand them out twice. If the same sled asks us for a share, because
+        /// it crashes or there is a network blip, we will return the same
+        /// share each time.
+        ///
+        /// Note that this is a fairly optimistic strategy as the requesting
+        /// sled can always go ask another sled after a network blip. However,
+        /// this guarantees that a single sled never hands out more than one of
+        /// its shares to any given sled.
+        ///
+        /// We can't do much better than this without some sort of centralized
+        /// distributor which is part of the reconfiguration mechanism in later
+        /// trust quourum protocol schemes.
+        distributed_shares: BTreeMap<Baseboard, ShareIdx>,
+    },
+    Learning,
+    Learned {
+        pkg: LearnedSharePkg,
+    },
+}
+
+impl State {
+    pub fn name(&self) -> &'static str {
+        match self {
+            State::Uninitialized => "uninitialized",
+            State::InitialMember { .. } => "initial_member",
+            State::Learning => "learning",
+            State::Learned { .. } => "learned",
+        }
+    }
+}
+
+/// A response to an Fsm API request
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum ApiOutput {
+    /// This peer has been initialized
+    ///
+    /// The caller *must* persist the state
+    PeerInitialized,
+
+    /// Rack initialization has completed. This node was the coordinator.
+    RackInitComplete,
+
+    /// A `RackSecret` was reconstructed
+    RackSecret { request_id: Uuid, secret: RackSecret },
+
+    /// An extra share has been distributed to a learning peer
+    ///
+    /// The caller *must* persist the state
+    ShareDistributedToLearner,
```

Comment:
It's confusing at first glance that this doesn't include any data, e.g. which share was distributed to which learner.  How do we know _what_ state to persist?

Reply:
We always persist the entire `Fsm::State`. What was distributed is updated there. This is just a marker to inform the caller to persist, similar to the old `Output::persist` boolean. The name is just more descriptive to help with logging etc... now.

Reply:
I changed the comments to explicitly say 
> The caller *must* persist `Fsm::State`

so that this would be clearer